### PR TITLE
[NOD-1032] Consensus updates pre-pruning

### PR DESCRIPTION
**Pull request by svarogg**

### CODE REVIEW

File: domain/blockdag/blocknode.go [hunk 1]
```diff
@@ -58,6 +77,9 @@ type blockNode struct {
 	// hundreds of thousands of these in memory, so a few extra bytes of
 	// padding adds up.
 
+	// dag is the blockDAG in which this node resides
+	dag *BlockDAG
```
Comment (stasatdaglabs) [inline domain/blockdag/blocknode.go:81]: The spaghetti thickens...

File: domain/blockdag/blocknode.go [hunk 2]
```diff
@@ -235,3 +266,111 @@ func (node blockNode) String() string {
...
+			return err
+		}
+
+		isRedInPastOfAnyNonFinalityViolatingBlue, err := node.dag.isInPastOfAny(red, nonFinalityViolatingBlues)
+		if err != nil {
+			return err
+		}
+
+		if !doesRedHaveFinalityPointInPast && !isRedInPastOfAnyNonFinalityViolatingBlue {
+			return ruleError(ErrViolatingObjectiveFinality, "block is violating objective finality")
```
Comment (stasatdaglabs) [inline domain/blockdag/blocknode.go:332]: Is "objective finality" the term we're using after all?

File: domain/blockdag/dag.go [hunk 3]
```diff
@@ -87,15 +86,19 @@ type BlockDAG struct {
...
 
 	reachabilityTree *reachabilityTree
 
 	recentBlockProcessingTimestamps []mstime.Time
 	startTime                       mstime.Time
+
+	tips blockSet
+
+	// validTips is a set of blocks with the status "valid", which have no valid descendants.
+	// Note that some validTips might not be actual tips.
```
Comment (stasatdaglabs) [inline domain/blockdag/dag.go:100]: Please write an essay somewhere explaining this. I will definitely forget otherwise.

File: domain/blockdag/dag.go [hunk 4]
```diff
@@ -443,5 +496,180 @@ func (dag *BlockDAG) IsKnownInvalid(hash *daghash.Hash) bool {
...
+func (dag *BlockDAG) selectVirtualParents(tips blockSet, finalityPoint *blockNode) (blockSet, error) {
+	selected := newBlockSet()
+
+	candidatesHeap := newDownHeap()
+	candidatesHeap.pushSet(tips)
+
+	// If the first candidate has been disqualified from the chain or violates finality -
+	// he cannot be virtual's parent, since it will make him virtual's selectedParent - disqualifying virtual itself.
+	// Therefore, in such a case we remove it from the list of virtual parent candidates, and replace with any of
+	// it's parents that have no disqualified children
```
Comment (stasatdaglabs) [inline domain/blockdag/dag.go:580]: it's -> its

**Review by stasatdaglabs — CHANGES_REQUESTED** [review]

File: domain/blockdag/validate.go [hunk 5]
```diff
@@ -964,115 +1034,245 @@ func validateCoinbaseMaturity(dagParams *dagconfig.Params, entry *UTXOEntry, txB
...
+		str := fmt.Sprintf("block contains " +
+			"transaction whose input sequence " +
+			"locks are not met")
+		return ruleError(ErrUnfinalizedTx, str)
+	}
+
+	return nil
+}
+
+func (dag *BlockDAG) checkTxOutputAmounts(tx *util.Tx, totalSompiIn uint64) (uint64, error) {
```
Comment (someone235) [inline domain/blockdag/validate.go:1139]: What's the difference between this and checkOutputsAmounts?

File: domain/blockdag/validate.go [hunk 6]
```diff
@@ -964,115 +1034,245 @@ func validateCoinbaseMaturity(dagParams *dagconfig.Params, entry *UTXOEntry, txB
...
+
+func (dag *BlockDAG) checkTxInputAmounts(
+	inputsWithReferencedUTXOEntries []*txInputAndReferencedUTXOEntry) (totalSompiIn uint64, err error) {
+
+	totalSompiIn = 0
+
+	for _, txInAndReferencedUTXOEntry := range inputsWithReferencedUTXOEntries {
+		utxoEntry := txInAndReferencedUTXOEntry.utxoEntry
+
+		// Ensure the transaction amounts are in range. Each of the
```
Comment (someone235) [inline domain/blockdag/validate.go:1166]: Why you don't use checkEntryAmounts?

File: domain/blockdag/validate.go [hunk 7]
```diff
@@ -1123,10 +1322,18 @@ func (dag *BlockDAG) CheckConnectBlockTemplateNoLock(block *util.Block) error {
 		return err
 	}
 
-	templateNode, _ := dag.newBlockNode(&header, dag.virtual.tips())
+	templateParents := newBlockSet()
+	for _, parentHash := range header.ParentHashes {
```
Comment (someone235) [inline domain/blockdag/validate.go:1326]: Don't we have LookupNodes for that?

File: domain/blockdag/validate.go [hunk 8]
```diff
@@ -1123,10 +1322,18 @@ func (dag *BlockDAG) CheckConnectBlockTemplateNoLock(block *util.Block) error {
...
+			return errors.Errorf("Couldn't find parent of templates with hash `%s`", parentHash)
+		}
+		templateParents.add(parent)
+	}
+
+	templateNode, _ := dag.newBlockNode(&header, templateParents)
 
-	_, err = dag.checkConnectToPastUTXO(templateNode,
-		dag.UTXOSet(), block.Transactions(), false)
+	err = dag.checkConnectBlockToPastUTXO(templateNode, dag.UTXOSet(), block.Transactions())
```
Comment (someone235) [inline domain/blockdag/validate.go:1336]: Can you just return it instead of returning the error one line after?

File: domain/blockdag/validate.go [hunk 9]
```diff
@@ -964,115 +1034,245 @@ func validateCoinbaseMaturity(dagParams *dagconfig.Params, entry *UTXOEntry, txB
...
 
-	if !fastAdd {
-		err := ensureNoDuplicateTx(pastUTXO, transactions)
 		if err != nil {
-			return nil, err
+			if ruleErr := &(RuleError{}); errors.As(err, ruleErr) {
+				return err
+			}
+
+			continue
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/validate.go:1053]: Good catch! 
I should have inverted the condition a couple of lines above - if this is not ruleError - return the error. If it is ruleError - ignore and continue.

File: domain/blockdag/validate.go [hunk 10]
```diff
@@ -964,115 +1034,245 @@ func validateCoinbaseMaturity(dagParams *dagconfig.Params, entry *UTXOEntry, txB
...
+
+	totalSompiOut, err := dag.checkTxOutputAmounts(tx, totalSompiIn)
 	if err != nil {
-		return nil, errors.Errorf("error getting bytes of fee data: %s", err)
+		return 0, 0, nil
 	}
 
-	if !fastAdd {
-		scriptFlags := txscript.ScriptNoFlags
+	txFee = totalSompiIn - totalSompiOut
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/validate.go:1104]: I disagree. I want this line to be close to where I calculate totalSompiIn and totalSompiOut, so that all related calculations are in the same area.

File: domain/blockdag/tips_test.go [hunk 11]
```diff
@@ -235,14 +249,11 @@ func TestChainUpdates(t *testing.T) {
 	// Create a chain to be removed
 	var toBeRemovedNodes []*blockNode
 	toBeRemovedTip := genesis
-	for i := 0; i < 5; i++ {
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/tips_test.go:238]: Added comments that explains how the test works, hope that's sufficient.

File: domain/blockdag/scriptval_test.go [hunk 12]
```diff
@@ -1,55 +1,46 @@
...
-		hash: blocks[0].Hash(),
-	}
-
-	scriptFlags := txscript.ScriptNoFlags
-	err = checkBlockScripts(node, utxoSet, blocks[0].Transactions(), scriptFlags, nil)
-	if err != nil {
-		t.Errorf("Transaction script validation failed: %v\n", err)
-		return
-	}
+	t.Skip() // TODO: Reactivate this test once we have blocks from testnet.
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/scriptval_test.go:10]: Done. Once you merge, please add to the ticket a link to the commit that removed the test, for reference.
Here's the ticket: https://daglabs.atlassian.net/browse/NOD-1299

File: domain/blockdag/process.go [hunk 13]
```diff
@@ -226,124 +228,225 @@ func (dag *BlockDAG) createBlockNodeFromBlock(block *util.Block) (
...
+	}
+
+	dbTx, err := dag.databaseContext.NewTx()
+	if err != nil {
+		return nil, err
+	}
+	defer dbTx.RollbackUnlessClosed()
+
+	if isNewSelectedTip {
+		err = dag.validateAndApplyUTXOSet(newNode, block, dbTx)
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/process.go:272]: applyDAGChanges can never return ruleError.
All it does is modify the DB etc.

Also note we have dbTx, just in case.

File: domain/blockdag/process.go [hunk 14]
```diff
@@ -226,124 +228,225 @@ func (dag *BlockDAG) createBlockNodeFromBlock(block *util.Block) (
...
+		if err != nil {
+			if !errors.As(err, &(RuleError{})) {
+				return nil, err
+			}
+			dag.index.SetBlockNodeStatus(newNode, statusDisqualifiedFromChain)
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/process.go:277]: I don't think this is newsworthy enough of a Warn log.

I mean, we don't care about block UTXO-validation enough to validate all blocks, but then if a block had the luck to get validated and wasn't found valid, we WARN about this? Seems a bit wrong for me.

Nevertheless, I agree that for debugging this would be useful.
Therefore, I've added to `log.Debugf("Accepted block blabla")` the status of the new block.
Note that to achieve that I had to move the log into maybeAcceptBlock, which is more then OK in my opinion.

File: domain/blockdag/process.go [hunk 15]
```diff
@@ -226,124 +228,225 @@ func (dag *BlockDAG) createBlockNodeFromBlock(block *util.Block) (
...
 
-	virtualUTXODiff, chainUpdates, err :=
-		dag.applyDAGChanges(node, newBlockPastUTXO, newBlockMultiSet, selectedParentAnticone)
+	err = dag.saveChangesFromBlock(block, dbTx)
 	if err != nil {
-		// Since all validation logic has already ran, if applyDAGChanges errors out,
-		// this means we have a problem in the internal structure of the DAG - a problem which is
-		// irrecoverable, and it would be a bad idea to attempt adding any more blocks to the DAG.
-		// Therefore - in such cases we panic.
-		panic(err)
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/process.go:263]: Because any error returned here will go all the way up to ProcessBlock, and then whomever calls ProcessBlock already knows that he should check for ruleError, and if it's not a ruleError - he should panic.

This is true without regards to where the error was returned, and therefore I see no reason why there should be a single place where we per-emptively panic.

File: domain/blockdag/process.go [hunk 16]
```diff
@@ -389,39 +480,71 @@ func (dag *BlockDAG) saveChangesFromBlock(block *util.Block, virtualUTXODiff *UT
...
-	// Allow the index manager to call each of the currently active
-	// optional indexes with the block being connected so they can
-	// update themselves accordingly.
-	if dag.indexManager != nil {
-		err := dag.indexManager.ConnectBlock(dbTx, block.Hash(), txsAcceptanceData)
+	return nil
+}
+
+func (dag *BlockDAG) boundedMergeBreakingParents(node *blockNode) (blockSet, error) {
+	potentiallyKosherizingBlocks, err := node.nonFinalityViolatingBlues()
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/process.go:487]: As much as I wish to take credit for the name, I'm afraid this is R's making.
Anyway, added a comment explain what this function does. Hope it's sufficient.

File: domain/blockdag/blocknode.go (hunk 1 above)
Comment (svarogg) [reply domain/blockdag/blocknode.go:81]: Is much better then passing node.blabla(dag) all the time.

File: domain/blockdag/blocknode.go (hunk 2 above)
Comment (svarogg) [reply domain/blockdag/blocknode.go:332]: Nope, changed to "Bounded merge depth" :stuck_out_tongue:

File: domain/blockdag/dag.go (hunk 3 above)
Comment (svarogg) [reply domain/blockdag/dag.go:100]: I'm not sure I understand what exactly you want to be written in said essay, and where do you want to see it?

File: domain/blockdag/block_utxo.go [hunk 17]
```diff
@@ -146,27 +140,69 @@ func (node *blockNode) applyBlueBlocks(selectedParentPastUTXO UTXOSet, blueBlock
...
+			return false, 0, 0, err
+		}
+
+		return true, 0, accumulatedMass, nil
+	}
+
+	txFee, accumulatedMassAfter, err = node.dag.checkConnectTransactionToPastUTXO(
+		node, tx, pastUTXO, accumulatedMassBefore, selectedParentMedianTime)
+	if err != nil {
+		if !errors.As(err, &(RuleError{})) {
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/block_utxo.go:189]: This will have to modify quite a few unrelated places, and I don't want to enhance this monster of PR any further, so created ticket:
https://daglabs.atlassian.net/browse/NOD-1366

File: domain/blockdag/validate.go [hunk 18]
```diff
@@ -964,115 +1034,245 @@ func validateCoinbaseMaturity(dagParams *dagconfig.Params, entry *UTXOEntry, txB
...
+
+	err = dag.checkTxSequenceLock(node, tx, inputsWithReferencedUTXOEntries, selectedParentMedianTime)
+	if err != nil {
+		return 0, 0, nil
+	}
+
+	err = ValidateTransactionScripts(tx, pastUTXO, txscript.ScriptNoFlags, dag.sigCache)
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/validate.go:1111]: Done that already ;)
https://daglabs.atlassian.net/browse/NOD-1358

File: domain/blockdag/blocknode.go [hunk 19]
```diff
@@ -235,3 +266,111 @@ func (node blockNode) String() string {
 func (node *blockNode) time() mstime.Time {
 	return mstime.UnixMilliseconds(node.timestamp)
 }
+
+func (node *blockNode) blockAtDepth(depth uint64) *blockNode {
+	if node.blueScore <= depth { // to prevent overflow of requiredBlueScore
+		depth = node.blueScore
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/blocknode.go:272]: That's what I did before, but this function is used when adding genesis to the dag, so `node.dag.gensis` is not set yet.

File: domain/blockdag/dag.go [hunk 20]
```diff
@@ -443,5 +496,182 @@ func (dag *BlockDAG) IsKnownInvalid(hash *daghash.Hash) bool {
...
+	dag.tips = newTips
+
+	return didVirtualParentsChange, chainUpdates, nil
+}
+
+func (dag *BlockDAG) updateVirtualParents(newTips blockSet, finalityPoint *blockNode) (
+	didVirtualParentsChange bool, chainUpdates *selectedParentChainUpdates, err error) {
+
+	var newVirtualParents blockSet
+	// If only genesis is the newTips - we are still initializing the DAG and not all structures required
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/dag.go:532]: What happens is that this function is called in setTips as `dag.updateVirtualParents(newTips, dag.virtual.finalityPoint())`.
However, if we are only inserting genesis rn, then dag.virtual is still not really initialized and dag.virtual.finalityPoint is therefore some bullshit block.

I've invested quite a lot of time to find a better architecture to this whole thing so this could be avoided, but couldn't find anything.
So I'll enhance the comment with what I wrote to you here, and if you have any ideas on how to make it better, please LMK (Or do yourself if I'm already on my vacation).

File: domain/blockdag/finality_conflicts.go [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+
+		isFinalityBlockInSelectedParentChain, err := dag.isInSelectedParentChainOf(finalityBlock, candidate)
+		if err != nil {
+			return err
+		}
+		if !isFinalityBlockInSelectedParentChain {
+			continue
+		}
+		if dag.index.BlockNodeStatus(candidate) == statusUTXONotVerified {
+			err = dag.resolveNodeStatus(candidate, dbTx)
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/finality_conflicts.go:71]: It was problematic because of the defer. Extracted the thing into new function that does everything including transaction. Hope you like.

File: domain/blockdag/process.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/process.go:277]: Pasting what I replied to Stas:

I don't think this is newsworthy enough of a Warn log.

I mean, we don't care about block UTXO-validation enough to validate all blocks, but then if a block had the luck to get validated and wasn't found valid, we WARN about this? Seems a bit wrong for me.

Nevertheless, I agree that for debugging this would be useful.
Therefore, I've added to log.Debugf("Accepted block blabla") the status of the new block.
Note that to achieve that I had to move the log into maybeAcceptBlock, which is more then OK in my opinion.

File: domain/blockdag/process.go [hunk 22]
```diff
@@ -226,124 +228,225 @@ func (dag *BlockDAG) createBlockNodeFromBlock(block *util.Block) (
...
 
-	// Update the virtual block's parents (the DAG tips) to include the new block.
-	chainUpdates = dag.virtual.AddTip(node)
+func (dag *BlockDAG) validateAndApplyUTXOSet(
+	node *blockNode, block *util.Block, dbTx *dbaccess.TxContext) error {
 
-	// Build a UTXO set for the new virtual block
-	newVirtualUTXO, _, _, err := dag.pastUTXO(&dag.virtual.blockNode)
+	if !node.isGenesis() {
+		err := dag.resolveNodeStatus(node.selectedParent, dbTx)
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/process.go:362]: We probably do. However, not sure will have time to deal with that, so created ticket:
https://daglabs.atlassian.net/browse/NOD-1369

File: domain/blockdag/process.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/process.go:277]: Because that's what makes sense.
validateAndApplyUTXOSet might also be called when resolving a finality conflict, and it should modify statuses.

File: domain/blockdag/process.go [hunk 23]
```diff
@@ -226,124 +228,225 @@ func (dag *BlockDAG) createBlockNodeFromBlock(block *util.Block) (
...
+		return nil, err
 	}
 
-	err = dag.saveChangesFromBlock(block, virtualUTXODiff, txsAcceptanceData, newBlockFeeData)
+	err = dbTx.Commit()
 	if err != nil {
 		return nil, err
 	}
 
+	dag.clearDirtyEntries()
```
Comment (svarogg) [inline domain/blockdag/process.go:313]: Because we can clearDirtyEntries only after we commited dbTx.
Now, dbTx is no longer the sole property of saveChangesFromBlock, so we can't commit dbTx inside it.

File: domain/blockdag/scriptval.go [hunk 24]
```diff
@@ -196,37 +195,3 @@ func ValidateTransactionScripts(tx *util.Tx, utxoSet UTXOSet, flags txscript.Scr
 	validator := newTxValidator(utxoSet, flags, sigCache)
 	return validator.Validate(txValItems)
 }
-
-// checkBlockScripts executes and validates the scripts for all transactions in
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/scriptval.go:200]: Why scriptval.go has more stuff in addition to that?
Or did you mean scriptval_test.go? In that case already removed due to Stas request ;)

File: domain/blockdag/tips_test.go [hunk 25]
```diff
@@ -30,16 +33,16 @@ func buildNode(t *testing.T, dag *BlockDAG, parents blockSet) *blockNode {
 	return nodeByMsgBlock(t, dag, block)
 }
 
-// TestVirtualBlock ensures that VirtualBlock works as expected.
-func TestVirtualBlock(t *testing.T) {
+// TestTips ensures that tips are updated as expected.
+func TestTips(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/tips_test.go:37]: A lot of things needs enhancement in tests.
I talked about this with Roman, he asked me to give half a day for tests (in which I wrote TestFinality), soon 3rd party guys are going to write a lot of tests for all blockdag.

File: domain/blockdag/tips_test.go [hunk 26]
```diff
@@ -235,14 +249,11 @@ func TestChainUpdates(t *testing.T) {
 	// Create a chain to be removed
 	var toBeRemovedNodes []*blockNode
 	toBeRemovedTip := genesis
-	for i := 0; i < 5; i++ {
+	for i := 0; i < 9; i++ {
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/tips_test.go:252]: See comment I wrote above the test. Hope it's sufficient.

File: domain/blockdag/utxodiffstore_test.go [hunk 27]
```diff
@@ -142,26 +142,4 @@ func TestClearOldEntries(t *testing.T) {
 			t.Fatalf("TestClearOldEntries: diffData for node %s is in the loaded set", node.hash)
 		}
 	}
-
-	// Add a block on top of the genesis to force the retrieval of all diffData
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/utxodiffstore_test.go:146]: No time to do it and not very urgent, so created a ticket:
https://daglabs.atlassian.net/browse/NOD-1371

File: domain/blockdag/validate.go [hunk 28]
```diff
@@ -824,6 +834,31 @@ func (dag *BlockDAG) checkBlockTransactionsFinalized(block *util.Block, node *bl
...
+	transactionsSet := make(map[daghash.TxID]struct{}, len(transactions))
+	for _, transaction := range transactions {
+		transactionsSet[*transaction.ID()] = struct{}{}
+	}
+
+	for _, transaction := range transactions {
+		for _, transactionInput := range transaction.MsgTx().TxIn {
+			if i, ok := transactionsSet[transactionInput.PreviousOutpoint.TxID]; ok {
+				str := fmt.Sprintf("block contains chained transactions: Input %d of transaction %s spend"+
+					"an output of transaction %s", i, transaction.ID(), transactionInput.PreviousOutpoint.TxID)
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/validate.go:853]: Yeah, i should have been assigned in the outer loop

File: domain/blockdag/validate.go [hunk 29]
```diff
@@ -879,14 +939,9 @@ func CheckTransactionInputsAndCalulateFee(tx *util.Tx, txBlueScore uint64, utxoS
...
 	for txInIndex, txIn := range tx.MsgTx().TxIn {
-		// Ensure the referenced input transaction is available.
-		entry, ok := utxoSet.Get(txIn.PreviousOutpoint)
-		if !ok {
-			str := fmt.Sprintf("output %s referenced from "+
-				"transaction %s input %d either does not exist or "+
-				"has already been spent", txIn.PreviousOutpoint,
-				tx.ID(), txInIndex)
-			return 0, ruleError(ErrMissingTxOut, str)
+		entry, err := checkReferencedOutputsAreAvailable(tx, utxoSet, txIn, txInIndex)
```
[the remark this answers is not in the document]
Comment (svarogg) [reply domain/blockdag/validate.go:942]: I'd rather not inline, so as to not clutter CheckTransactionInputsAndCalulateFee, but renamed to findReferencedOutput.
Hope this helps.

File: domain/blockdag/process.go [hunk 30]
```diff
@@ -483,8 +497,13 @@ func (dag *BlockDAG) saveChangesFromBlock(block *util.Block, dbTx *dbaccess.TxCo
 	return nil
 }
 
+// boundedMergeBreakingParents returns all parents of given `node` that break the bounded merge depth rule:
+// All blocks in B.MergeSet should be in future of node.finalityPoint.
+// If there exists a block C violating this, i.e., C in node's merge set and node.finalityPoint's anticone,
+// then there must be a (“kosherizing”) block D in C's Future such that D is in node.blues
+// and node.finalityPoint in D.SelectedChain; and
```
Comment (someone235) [inline domain/blockdag/process.go:504]: This is cut in the middle

File: domain/blockdag/tips_test.go [hunk 31]
```diff
@@ -232,6 +220,12 @@ func TestSelectedPath(t *testing.T) {
 	virtual2.updateSelectedParentSet(buildNode(t, dag, blockSetFromSlice()))
 }
 
+// TestChainUpdates makes sure the chainUpdates from setTips are correct:
+// It creates two chains: a main-chain to be removed and a side-chain to be added
+// The main-chain has to be longer then the side-chain, so that the natural selected tip of the DAG is the one
```
Comment (someone235) [inline domain/blockdag/tips_test.go:225]: then->than

File: domain/blockdag/validate.go [hunk 32]
```diff
@@ -936,10 +936,9 @@ func CheckTransactionInputsAndCalulateFee(
 		return 0, nil
 	}
 
-	txID := tx.ID()
 	var totalSompiIn uint64
 	for txInIndex, txIn := range tx.MsgTx().TxIn {
-		entry, err := checkReferencedOutputsAreAvailable(tx, utxoSet, txIn, txInIndex)
+		entry, err := findReferencedOutput(tx, utxoSet, txIn, txInIndex)
```
Comment (someone235) [inline domain/blockdag/validate.go:941]: I still don't like the fact that the only use of findReferencedOutput is to wrap the error behind utxoSet.Get(). Don't you prefer to just inline it?

File: domain/blockdag/error.go [hunk 33]
```diff
@@ -203,13 +203,30 @@ const (
...
 
 	// ErrOrphanBlockIsNotAllowed indicates that an orphan block was submitted with
 	// BFDisallowOrphans flag raised.
 	ErrOrphanBlockIsNotAllowed
+
+	// ErrViolatingObjectiveFinality indicates that a block is violating finality from
+	// it's own point of view
+	ErrViolatingObjectiveFinality
+
+	// ErrViolatingMergeLimit indicates that a block merges more then mergeLimit blocks
```
[the remark this answers is not in the document]
Comment (someone235) [reply domain/blockdag/error.go:221]: This is not fixed

**Review by someone235 — CHANGES_REQUESTED** [review]

File: domain/blockdag/validate.go (hunk 32 above)
Comment (svarogg) [reply domain/blockdag/validate.go:941]: The same is more or less true for validateCoinbaseMaturity, checkEntryAmounts, etc.
But this makes CheckTransactionInputsAndCalulateFee much more readable, where you can with a single glance understand all that it does.

File: domain/blockdag/process.go [hunk 34]
```diff
@@ -501,7 +501,7 @@ func (dag *BlockDAG) saveChangesFromBlock(block *util.Block, dbTx *dbaccess.TxCo
 // All blocks in B.MergeSet should be in future of node.finalityPoint.
 // If there exists a block C violating this, i.e., C in node's merge set and node.finalityPoint's anticone,
 // then there must be a (“kosherizing”) block D in C's Future such that D is in node.blues
```
Comment (someone235) [inline domain/blockdag/process.go:505]: Why do you put parenthesis?

File: domain/blockdag/process.go [hunk 35]
```diff
@@ -501,7 +501,7 @@ func (dag *BlockDAG) saveChangesFromBlock(block *util.Block, dbTx *dbaccess.TxCo
 // All blocks in B.MergeSet should be in future of node.finalityPoint.
```
Comment (someone235) [inline domain/blockdag/process.go:503]: `All blocks in B.MergeSet should be in future of node.finalityPoint` This sentence is false. Please rephrase it to be something like `All blocks in B.MergeSet should be in future of node.finalityPoint except...' Or 'All blocks in B.MergeSet should be in future of node.finalityPoint, with the following exception:'

**Review by someone235 — CHANGES_REQUESTED** [review]

**Review by someone235 — DISMISSED** [review]

File: domain/blockdag/dag.go (hunk 3 above)
Comment (stasatdaglabs) [reply domain/blockdag/dag.go:100]:
```Tips - actual tips
ValidTips - tips of the valid dag (a.k.a. all blocks that have no statusValid descendants. A.k.a. those that have virtual as their DiffChild)
Virtual parents - parents of virtual. Have some complicated rules to determine them (see selectVirtualParents)```

File: domain/blockdag/blocknode.go [hunk 36]
```diff
@@ -68,9 +92,12 @@ type blockNode struct {
 	// children are all the blocks that refer to this block as a parent
 	children blockSet
 
-	// blues are all blue blocks in this block's worldview that are in its selected parent anticone
+	// blues are all blue blocks in this block's worldview that are in its merge set
 	blues []*blockNode
 
+	// reds are all red blocks in this block's worldview that are in its its merge set
```
Comment (stasatdaglabs) [inline domain/blockdag/blocknode.go:98]: "its its" -> "its"

File: domain/blockdag/blocknode.go [hunk 37]
```diff
@@ -235,3 +268,111 @@ func (node blockNode) String() string {
...
+	return node.blockAtDepth(node.dag.FinalityInterval())
+}
+
+func (node *blockNode) hasFinalityPointInOthersSelectedChain(other *blockNode) (bool, error) {
+	finalityPoint := node.finalityPoint()
+	return node.dag.isInSelectedParentChainOf(finalityPoint, other)
+}
+
+func (node *blockNode) nonBoundedMergeDepthViolatingBlues() (blockSet, error) {
+	nonFinalityViolatingBlues := newBlockSet()
```
Comment (stasatdaglabs) [inline domain/blockdag/blocknode.go:300]: Please update this variable name.

File: domain/blockdag/process.go [hunk 38]
```diff
@@ -389,39 +496,78 @@ func (dag *BlockDAG) saveChangesFromBlock(block *util.Block, virtualUTXODiff *UT
...
-	if dag.indexManager != nil {
-		err := dag.indexManager.ConnectBlock(dbTx, block.Hash(), txsAcceptanceData)
+	return nil
+}
+
+// boundedMergeBreakingParents returns all parents of given `node` that break the bounded merge depth rule:
+// All blocks in B.MergeSet should be in future of node.finalityPoint, with the following exception:
+// If there exists a block C violating this, i.e., C in node's merge set and node.finalityPoint's anticone,
+// then there must be a "kosherizing" block D in C's Future such that D is in node.blues
+// and node.finalityPoint in D.SelectedChain
```
Comment (stasatdaglabs) [inline domain/blockdag/process.go:506]: Please define B, C, and D.

**Review by stasatdaglabs — CHANGES_REQUESTED** [review]

**Review by someone235 — APPROVED** [review]

**Review by stasatdaglabs — APPROVED** [review]

File: app/rpc/rpccontext/notificationmanager.go [hunk 39]
```diff
@@ -19,12 +19,22 @@ type OnBlockAddedListener func(notification *appmessage.BlockAddedNotificationMe
 // OnChainChangedListener is a listener function for when the DAG's selected parent chain changes
 type OnChainChangedListener func(notification *appmessage.ChainChangedNotificationMessage) error
 
+// OnFinalityConflictListener is a listener function for when there a finality conflict in the DAG
```
Comment (someone235) [inline app/rpc/rpccontext/notificationmanager.go:22]: there->there's

**Review by someone235 — CHANGES_REQUESTED** [review]

**Review by someone235 — APPROVED** [review]

**Review by someone235 — APPROVED** [review]

**Review by stasatdaglabs — APPROVED** [review]

File: domain/blockdag/scriptval_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (elichai) [reply domain/blockdag/scriptval_test.go]: :/

[the remark this answers is not in the document]
Comment (elichai) [reply domain/blockdag/scriptval_test.go]: :/ no commit was added
